chained/tlsfrag.go

```diff
@@ -32,20 +34,41 @@ func makeFragFunc(pc *config.ProxyConfig) (tlsfrag.FragFunc, bool) {
 	}
 
 	// fragStr should be of the form <frag func> or <frag func>:<func config>.
-	// Example: "index:3"
+	// Example: "index:3" or "regex:foo.*" or "indexOf:test"
 	funcType, cfg, hasCfg := strings.Cut(fragStr, ":")
+	if !hasCfg {
+		log.Error("tlsfrag: missing config specifier")
+		return nil, false
+	}
 	switch funcType {
 	case "index":
-		if !hasCfg {
-			log.Error("tlsfrag: missing index specifier")
-			return nil, false
-		}
-
 		index, err := strconv.Atoi(cfg)
 		if err != nil {
 			log.Errorf("tlsfrag: bad index specifier: %v", err)
+			return nil, false
 		}
 		return func(_ []byte) int { return index }, true
+	case "regex":
+		regex, err := regexp.Compile(cfg)
+		if err != nil {
+			log.Errorf("tlsfrag: bad regex specifier: %v", err)
+			return nil, false
+		}
+		return func(b []byte) int {
+			loc := regex.FindIndex(b)
```

Comment:
I'm not totally sure that this works.  It sounds like the `regexp` package expects the input to be UTF-8 code-points.  I'm not sure how that would behave with a ClientHello.  I think the text, like the value of the SNI might be UTF-8 encoded.  But in theory there could be unexpected collisions with the binary data.  I guess that's unlikely?  But technically the behavior would be undefined.  It'd also be difficult to look for stuff like the SNI extension without knowing the SNI ahead of time.  You could look for the expected UTF-8 code-points, but that's pretty awkward IMO.

It's a bummer because specifying a split function as regex would be super powerful.  I thought it was an awesome idea.

Reply:
Oh, what if the regular expression expects to see the ClientHello as a hex-encoded string?  To be concrete, here you would have:
```
return func(b []byte) int {
    loc := regex.FindStringIndex(hex.EncodeToString(b))
    // handle loc == nil
    return loc[0]
}
```

Reply:
Interesting yeah I like the hex encoding idea, although I think it would throw the index off? Maybe we could figure out what the index should be though?

At least in that test with the test hello, the straight regex works, but I agree the regex theoretically is looking for UTF-8. I thought it might bork at non-UTF-8 bytes, but that doesn't appear to be true in the test at least.